charmcraft/commands/store/registry.py

```diff
@@ -249,3 +279,34 @@ def get_destination_url(self, reference):
         _, digest, _ = self.registry.get_manifest(reference)
         final_fqu = self.registry.get_fully_qualified_url(digest)
         return final_fqu
+
+    def _extract_file(self, image_tar, name, compress=False):
+        """Extract a file from the tar and return its info. Optionally, gzip the content."""
+        logger.debug("Extracting file %r from local tar (compress=%s)", name, compress)
+        src_fh = image_tar.extractfile(name)
+        mtime = image_tar.getmember(name).mtime
+
+        hashing_temp_file = HashingTemporaryFile()
+        if compress:
+            # open the gzip file using the temporary file handler; use the original name and time
+            # as 'filename' and 'mtime' correspondingly as those go to the gzip headers,
+            # to ensure same final hash across different runs
+            dst_filehandler = gzip.GzipFile(
+                fileobj=hashing_temp_file,
+                mode="wb",
+                filename=os.path.basename(name),
+                mtime=mtime,
+            )
+        else:
+            dst_filehandler = hashing_temp_file
+        while True:
+            chunk = src_fh.read(CHUNK_SIZE)
+            if not chunk:
+                break
+            dst_filehandler.write(chunk)
+        dst_filehandler.close()
```

Comment:
do you intend to catch any error produced by this and retry? If so, this may warrant a try and finally to ensure the close takes place for this and the temp file.

Also, aren't both of these handles? The naming is different for each

Reply:
I don't intend to catch errors here, this is a local communication that actually ends quite fast, I don't think we won't see many failures here.

Improved the naming.

Reply:
Ideally files should be closed on write error, but I don't know the usage context to see if it could lead to a descriptor leak or something similar. From your previous comment I assume that there will be no major consequences of leaving them open, but adding a try-finally around the while loop shouldn't cause any problems.

Reply:
Improved, thanks.